### PR TITLE
regression tests for pointer invalidation in core library slice methods

### PR DESCRIPTION
**Pull request by Pointerbender**
A fix for a pointer invalidation bug in `<[T]>::copy_within` has [landed](https://github.com/rust-lang/rust/pull/85610) on the Rust master branch. This PR updates the `rust-version` file to the latest master commit hash and adds extra tests to the Miri test suite to ensure that regressions of this type of bug can be detected for various slice methods with the `-Zmiri-track-raw-pointers` flag.

I took the liberty of adding 2 extra  `#![feature]` attributes at the top of `slices.rs`, since there already was one unstable feature. I hope this is okay :smile: 

One thing I noticed when running the entire Miri test suite with `MIRIFLAGS="-Zmiri-track-raw-pointers" ./miri test` is that there are currently failing tests on the master branch:

```
failures:
    [ui] run-pass/align.rs
    [ui] run-pass/box.rs
    [ui] run-pass/concurrency/simple.rs
    [ui] run-pass/libc.rs
    [ui] run-pass/ptr_int_casts.rs
    [ui] run-pass/stacked-borrows/int-to-ptr.rs

test result: FAILED. 199 passed; 6 failed; 1 ignored; 0 measured; 0 filtered out; finished in 12.95s
```

These failures were not fixed in this PR and already existed prior to this PR. I haven't investigated these yet, but am interested in helping out if possible!

Thanks!

### CODE REVIEW

**Comment by RalfJung** [comment]
> One thing I noticed when running the entire Miri test suite with MIRIFLAGS="-Zmiri-track-raw-pointers" ./miri test is that there are currently failing tests on the master branch:

Yeah, it is expected that some tests fail. In particular, those that cast integers to pointers cannot work with this flag. That will explain some of the test failures, though possibly not all of them.

File: tests/run-pass/slices.rs [hunk 1]
```diff
@@ -186,8 +188,60 @@ fn uninit_slice() {
...
+
+    unsafe { assert_eq!(0, *buffer.as_mut_ptr_range().start ); }
+    // Check that the pointer range is in-bounds, while we're at it
+    let range = buffer.as_mut_ptr_range();
+    unsafe { assert_eq!(*range.start, *range.end.sub(len)); }
+
+    buffer.reverse();
+
+    // Calls `fn as_chunks_unchecked_mut` internally (requires unstable `#![feature(slice_as_chunks)]`):
+    assert_eq!(2, buffer.as_chunks_mut::<32>().0.len());
```
Comment (RalfJung) [inline tests/run-pass/slices.rs:222]: Some of these aliasing errors only arise when you actually use the resulting slice for something. So maybe add a loop like this?
```rust
for chunk in &mut buffer.as_chunks_mut::<32>().0 {
  for elem in chunk {
    *elem += 1;
  }
}
```

Same for `partition_dedup` below.

**Comment by Pointerbender** [comment]
Thanks for the PR comments @RalfJung! I will address these.

> Sure, if you want to audit these and make sure they are all due to int-to-ptr casts, that would be great. :)
> Let's do that separately from this PR though.

Sounds good! In case the failures are all due to int-to-ptr casts, would it be helpful if I can rewrite the test cases without casting pointers to integers, while keeping the same functionality?

Thanks!

p.s. I'm going on a short trip tomorrow, so it may take me until after the weekend to get back with more updates.

**Comment by RalfJung** [comment]
> Sounds good! In case the failures are all due to int-to-ptr casts, would it be helpful if I can rewrite the test cases without casting pointers to integers, while keeping the same functionality?

I think (almost) all of these tests are explicitly about testing int-to-ptr casts, so changing them would defeat their purpose. We do want to continue supporting int-to-ptr casts, and there is a reason that `-Zmiri-track-raw-pointers` is not the default.

> p.s. I'm going on a short trip tomorrow, so it may take me until after the weekend to get back with more updates.

Sure, no worries.

**Comment by Pointerbender** [comment]
Hi @RalfJung, thank you for your patience while I was away! I have just pushed a new version which resolves the PR comments. I also rebased with the master branch upstream, due to a merge conflict in the `rust-version` file. I think everything should be good to go, but do let know if you encounter anything else :smiley: Thanks!

**Comment by RalfJung** [comment]
Looking great, thanks a lot @Pointerbender :)
@bors r+

**Pull request merged**